examples/vae/train.py

```diff
@@ -145,59 +70,41 @@ def eval_model(vae):
     metrics = compute_metrics(recon_images, images, mean, logvar)
     return metrics, comparison, generate_images
 
-  return nn.apply(eval_model, model())({'params': params})
-
-
-def prepare_image(x):
-  x = tf.cast(x['image'], tf.float32)
-  x = tf.reshape(x, (-1,))
-  return x
-
+  return nn.apply(eval_model, models.model(flags))({'params': params})
 
-def main(argv):
-  del argv
-
-  # Make sure tf does not allocate gpu memory.
-  tf.config.experimental.set_visible_devices([], 'GPU')
 
+def train_and_evaluate(flags):
+  """Train and evaulate pipeline."""
   rng = random.PRNGKey(0)
   rng, key = random.split(rng)
 
-  ds_builder = tfds.builder('binarized_mnist')
-  ds_builder.download_and_prepare()
-  train_ds = ds_builder.as_dataset(split=tfds.Split.TRAIN)
-  train_ds = train_ds.map(prepare_image)
-  train_ds = train_ds.cache()
-  train_ds = train_ds.repeat()
-  train_ds = train_ds.shuffle(50000)
-  train_ds = train_ds.batch(FLAGS.batch_size)
-  train_ds = iter(tfds.as_numpy(train_ds))
-
-  test_ds = ds_builder.as_dataset(split=tfds.Split.TEST)
-  test_ds = test_ds.map(prepare_image).batch(10000)
-  test_ds = np.array(list(test_ds)[0])
-  test_ds = jax.device_put(test_ds)
+  logging.info('Initializing dataset.')
+  train_ds = input_pipeline.build_train_set(flags)
+  test_ds = input_pipeline.build_test_set()
 
-  init_data = jnp.ones((FLAGS.batch_size, 784), jnp.float32)
+  logging.info('Initializing model.')
+  init_data = jnp.ones((flags.batch_size, 784), jnp.float32)
+  params = models.model(flags).init(key, init_data, rng)['params']
 
   state = train_state.TrainState.create(
-      apply_fn=model().apply,
-      params=model().init(key, init_data, rng)['params'],
-      tx=optax.adam(FLAGS.learning_rate),
+      apply_fn=models.model(flags).apply,
+      params=params,
+      tx=optax.adam(flags.learning_rate),
   )
 
   rng, z_key, eval_rng = random.split(rng, 3)
-  z = random.normal(z_key, (64, FLAGS.latents))
+  z = random.normal(z_key, (64, flags.latents))
 
-  steps_per_epoch = 50000 // FLAGS.batch_size
+  steps_per_epoch = 50000 // flags.batch_size
```

Comment:
when you create the datasetbuilder in this file (see my other comment in `input_pipeline.py`) then you can read the number of examples from `builder.info` instead of hard-coding it here.